File: linter.rb
class Linter

  attr_reader :error

  def initialize
    @stack = []
  end

  def lint(text)
    text.each_char.with_index do |char, index|

      if opening_brace?(char)
        @stack.push(char)
      elsif closing_brace?(char)
        if closes_most_recent_opening_brace?(char)
          @stack.pop
        else
          @error = "Incorrect closing brace: #{char} at index #{index}"
          return
        end
      end
    end

    if @stack.any?
      @error = '#{@stack.last}'
    end
  end

  private

  def opening_brace?(char)
    ["(", "[", "{"].include?(char)
  end

  def closing_brace?(char)
    [")", "]", "}"].include?(char)
  end

  def most_recent_opening_brace
    @stack.last
  end

  def closes_most_recent_opening_brace?(char)
    opening_brace_of(char) == most_recent_opening_brace
  end
end

linter = Linter.new
linter.lint("(var x = { y: [1, 2, 3] })")

File: linter.js
class Linter {
    constructor() {
        this.stack = [];
        this.error = undefined;
    }

    lint(text) {
        if(typeof text !== "string") {
            console.log("Wrong args provided to linter");
            return;
        }

        let charArr = text.split(""),
            length = charArr.length;

        for(let i = 0; i < length; i++) {

            let nextChar = charArr[i];

            if(this._openingBrace(nextChar)) {
                this.stack.push(nextChar);
            } else if(this._closingBrace(nextChar)) {
                if(this._closesMostRecentBrace(nextChar)) {
                    this.stack.pop();
                } else {
                    this.error = `Incorrect closing brace ${nextChar} at index ${i}`;
                    console.log(this.error);
                    return;
                }

            }
        }

        if(this.stack.length) {
            this.error = `Stack last ${this.stack}`
        }
    }

    _openingBrace(char) {
        return ["(", "[", "{"].includes(char);
    }

    _closingBrace(char) {
        return ["]", "}", ")"].includes(char);
    }

    _mostRecentBraceInStack() {
        return this.stack[-1];
    }

    _openingBraceOfCurrentBrace(char) {
        const braceMap = {
            "}" : "{",
            ")" : "(",
            "]" : "["
        };

        return braceMap[char];
    }

    _closesMostRecentBrace(char) {
        return this._openingBraceOfCurrentBrace(char) === this._mostRecentBraceInStack();
    }
}

const linter = new Linter();
linter.lint("(var x = { y: [1, 2, 3] })");